src/Datadog.Trace.DuckTyping/ILHelpersExtensions.cs

```diff
@@ -11,23 +11,52 @@ namespace Datadog.Trace.DuckTyping
     /// </summary>
     internal static class ILHelpersExtensions
     {
-        private static Func<DynamicMethod, RuntimeMethodHandle> _dynamicGetMethodDescriptor;
-        private static List<RuntimeMethodHandle> _handles = new List<RuntimeMethodHandle>();
+        private static List<DynamicMethod> _dynamicMethods = new List<DynamicMethod>();
 
-        static ILHelpersExtensions()
+        internal static DynamicMethod GetDynamicMethodForIndex(int index)
         {
-            _dynamicGetMethodDescriptor = (Func<DynamicMethod, RuntimeMethodHandle>)typeof(DynamicMethod)
-                .GetMethod("GetMethodDescriptor", BindingFlags.NonPublic | BindingFlags.Instance)
-                .CreateDelegate(typeof(Func<DynamicMethod, RuntimeMethodHandle>));
+            lock (_dynamicMethods)
+            {
+                return _dynamicMethods[index];
+            }
+        }
+
+        internal static void CreateDelegateTypeFor(TypeBuilder proxyType, DynamicMethod dynamicMethod, out Type delType, out MethodInfo invokeMethod)
+        {
+            ModuleBuilder modBuilder = (ModuleBuilder)proxyType.Module;
+            TypeBuilder delegateType = modBuilder.DefineType($"{dynamicMethod.Name}Delegate_" + Guid.NewGuid().ToString("N"), TypeAttributes.Class | TypeAttributes.Public | TypeAttributes.Sealed | TypeAttributes.AnsiClass | TypeAttributes.AutoClass, typeof(MulticastDelegate));
```

Comment:
Is there a cheap way to have a name that guarantees uniqueness? If not I'm OK with guid

Reply:
I think maybe we can have a global counter of the delegates already emitted, that should be enough don't you think?